File: docs/ConvNet/ConvNet.js

function ConvNet(){

  this.layers = [];

  this.addLayer = function(type, param1, param2){
    var layer = {type: type};

    if(type=='conv'){
      layer.filters = this.createFilters(param1, param2);
      layer.width = param2;
    }else if(type=='pool'){
      layer.width = param1; //!!! : width = 2 dans pooling
      // ?:layer.stride =
      layer.take = param2;
    }else if(type=='relu'){
      layer.min = param1;
    }

    this.layers.push(layer);
  };

  this.forward = function(img){

    //TODO: vérifier si 'image' est une seule image ou un tableau d'image

    /*var data = {
      images: [image]
    };*/

    var images = [img];

    for(var l=0; l<this.layers.length; l++){
      var layer = this.layers[l];
      var nImages = images.length;  //Le nombre d'images peut changer

      console.log('layer', l, '-', layer.type);

      if(layer.type != 'fc'){

        for(var i=0; i<nImages; i++){
          var output = this.goToNextLayer(layer, images[i]);

          if(layer.type=='conv'){  //'conv' renvoie un tableau d'image
            images[i] = output[0];
            for(var e=1; e<output.length; e++) images.push(output[e]);
          }else images[i] = output;
        }

      }else{
        var newImage = [];

        for(var i=0; i<nImages; i++){
          for(var value=0; value<images[i].length; value++){
            newImage.push(images[i][value]);
          }

        }

        images = newImage;

      }
    }

    return images;
  };


  this.goToNextLayer = function(layer, image){
    switch(layer.type){
      case 'conv':
        //console.log(layer);
        return convolution(layer, image);

      case 'pool':
        return pooling(layer, image);

      case 'relu':
        return relu(layer, image);

      case 'fc':
        return fc(layer,image);

      default:
        console.log('error !!! : no layer');
    }

    return image;
  };


  this.createFilters = function(n, w){
    var filters = [];
    for(var f=0; f<n; f++){
      var filter = [];
      for(var i=0; i<w*w; i++){
        var r = getRandom(-1,1);
        //if(Math.abs(r)<1) r = Math.abs(r)+1;
        filter.push(r);
      }
      filters.push(filter);
    }
    return filters;
  };

}

function getRandom(min, max){  //Nombre aléatoire entre deux valeurs
  return Math.random()*(max-min) + min;
}

var conv = {
  type: 'conv',
  width: 3,
  filters : [ [] ]
};


function convolution(conv, image){

  var imageFilter = { //Rassemble les données pour le processus
    image: image,
    iWidth: Math.sqrt(image.length),  //Largeur de l'image
    fWidth: conv.width,  //Largeur du filtre
    padding: (conv.width-1)/2
  };

  var images = [];
  for(var f=0; f<conv.filters.length; f++){  //Pour chaque filtre:
    var filter = conv.filters[f];
    imageFilter.filter = filter;
    imageFilter.sumFilter = sumArray(filter);
    //console.log(imageFilter);
    var newImage = imageConv(imageFilter);
    images.push(newImage);
  }
  return images;
}


function imageConv(imageFilter){  // Calcul de l'image filtrée

  var iWidth = imageFilter.iWidth;
  var padding = imageFilter.padding;

  var newImage = [];
  //arrayCopy(image, newImage); //Récupère les dimensions de 'image'

  for(var y=padding; y<iWidth-padding; y++){
    for(var x=padding; x<iWidth-padding; x++){
      newImage.push(filterValue(x, y, imageFilter));
    }
  }

  return newImage;
}

function filterValue(x, y, imageFilter){  //Calcul de la valeur d'un pixel
  var image = imageFilter.image;
  var filter = imageFilter.filter;
  var sumFilter = imageFilter.sumFilter;

  var iWidth = imageFilter.iWidth;
  var fWidth = imageFilter.fWidth;
  var padding = imageFilter.padding;

  var sum = 0;
  for(var i=-padding; i<=padding; i++){
    for(var j=-padding; j<=padding; j++){
      var index = (x+j) + (y+i)*iWidth;
      var fIndex = (j+padding) + (i+padding)*fWidth
      sum += image[index] * filter[fIndex];
    }
  }

  //var sumFilter = filter.reduce((pv, cv) => Math.abs(pv)+Math.abs(cv), 0);  //Somme des poids du filtre
  sum /= sumFilter;

  // TODO : Vérifier s'il est nécessaire de contraindre les valeurs
  if(sum<0) sum = 0;
  if(sum>255) sum = 255;

  return sum;
}

function sumArray(array){
  var sum=0;
  for(var i=0; i<array.length; i++){
    sum+=array[i];
  }
  return sum;
}



function pooling(pool, image){
  var width = Math.sqrt(image.length);
  var pImage = {
    image: image,
    width: width
  };
  var newImage = [];
  for(var y=0; y<width; y+=2){
    for(var x=0; x<width; x+=2){
      newImage.push(pooli(x, y, pImage));
    }
  }
  return newImage;
}

function pooli(x, y, pImage){
  var image = pImage.image;
  var width = pImage.width;
  var max = image[x + y*width];                   //Haut-Gauche
  max = Math.max(max, image[x+1 + y*width]);      //Haut-Droite
  max = Math.max(max, image[x +(y+1)*width]);     //Bas-Gauche
  max = Math.max(max, image[x+1 + (y+1)*width]);  //Bas-Droite
  return max;
}


function relu(relu, image){
  var min = relu.min;

  for(var i=0; i<image.length; i++){
    if(image[i]<min) image[i]=min;
  }

  return image;
}

function fc(fc, image){


}

File: Server/ConvNet.py

from math import sqrt
from random import uniform

import sys


class ConvNet:

    def __init__(self):
        self.layers = []


    def addLayer(self, t, arg1, arg2 = 3):
        layer = {}
        layer['type'] = t;
        # ou layer = {'type: t'}

        if t == 'conv':
            layer['filters'] = self.createFilters(arg1, arg2)
            layer['width'] = arg2
        elif t == 'pool':
            layer['width'] = arg2
            #layer['stride'] =
        elif t == 'relu':
            layer['min'] = arg1

        self.layers.append(layer)

    def forward(self, img):
        images =  [] #Liste des images créées par convulation
        images.append(img)

        #print(images)

        for layer in self.layers:

            print('layer - ' + layer['type']);

            for i in range(len(images)):
                #Récupère l'image:
                image = images[i]

                #Sortie de la couche :
                output = self.layerOutput(layer, image)

                if layer['type'] == 'conv':   #'conv' renvoie un tableau d'image
                    images[i] = output[0]   #Remplace l'image utilisée
                    for e in range(1,len(output)):
                        images.append(output[e])
                else:
                    images[i] = output

        return images


    def layerOutput(self, layer, image):  # Redirection vers la fonction de la couche
        t = layer['type']

        if t == 'conv':
            return convolution(layer, image)
        if t == 'pool':
            return pooling(layer, image)
        if t == 'relu':
            return relu(layer, image)

        return image

    def createFilters(self, n, width):  # Créé 'n' filtres de dimensions 'width' par 'width' (1 filtre = suite de nombres)
        filtres = []    #Listes des filtres

        for f in range(n):
            filtre = []
            for i in range(width*width):
                filtre.append(0.5)#uniform(-1,1))

            filtres.append(filtre)

        print(filtres)
        return filtres


#Couches:
#Layer
def convolution(conv, image):
    imageFilter = {
        'image': image,
        'iWidth': int(sqrt(len(image))),
        'fWidth': conv['width'],
        'padding': int((conv['width']-1)/2)
    }

    images = []
    for filtre in conv['filters']:
        imageFilter['filter'] = filtre
        imageFilter['sumFilter'] = sum(filtre)
        newImage = imageConv(imageFilter)
        images.append(newImage)

    return images

def imageConv(imageFilter):
    iWidth = imageFilter['iWidth']
    padding = imageFilter['padding']

    newImage = []

    for y in range(padding, iWidth-padding):
        for x in range(padding, iWidth-padding):
            newImage.append(filterValue(x, y, imageFilter))

    return newImage

def filterValue(x, y, imageFilter):
    image = imageFilter['image']
    filtre = imageFilter['filter']
    sumFilter = imageFilter['sumFilter']
    iWidth = imageFilter['iWidth']
    fWidth = imageFilter['fWidth']
    padding = imageFilter['padding']

    s=0
    for i in range(-padding, padding+1):
        for j in range(-padding, padding+1):
            index = (x+j) + (y+i)*iWidth
            fIndex = (j+padding) + (i+padding)*fWidth
            s += image[index] * filtre[fIndex]

    s /= sumFilter

    # TODO : Vérifier s'il est nécessaire de contraindre les valeurs
    if s<0: s=0
    if s>255: s=255

    return s

#Layer
def pooling(pool, image):
    width = int(sqrt(len(image)))
    pImage = {'image': image, 'width': width}
    newImage = []

    for y in range(0, width, 2):
        for x in range(0, width, 2):
            newImage.append(pooli(x,y,pImage))

    return newImage

def pooli(x,y,pImage):  #Retourne le maximum dans une zone de l'image
    image = pImage['image']
    width = pImage['width']
    m = image[x + y*width]                  #Haut-Gauche
    m = max(m, image[x+1 + y*width])        #Haut-Droite
    m = max(m, image[x + (y+1)*width])      #Bas-Gauche
    m = max(m, image[x+1 + (y+1)*width])    #Bas-Droite
    return m

#Layer
def relu(relu, image):
    m = relu['min']
    for i in range(len(image)):
        if image[i] < m:
            image[i]=m
    return image



"""
net = ConvNet()

net.addLayer('conv',3,3)
#net.addLayer('relu',8)
#net.addLayer('pool',2)

image = [2,0,2,6,4,8,5,4,2,12,8,45,12,54,5,4]


print(net.forward(image))

"""
numbers = []
s = sys.argv[1]

numbers.append(s)
numbers.append(s)

print(numbers)
sys.stdout.flush()
